text/0000-portability-lint.md

```diff
@@ -0,0 +1,574 @@
+built-in portabilities, which are fed to the SAT solver when comparing
+portabilities. The end result is that code like the following should pass the lint:
+
+```rust
+#[cfg(unix)]
+fn unix_only() { .. }
+
+#[cfg(linux)]
+fn linux_only() {
+    // permitted since `linux` implies `unix`
+    unix_only()
+}
+```
+
+### Determining the portability of referenced items
+
+**How is the portability of a referenced item determined?** The lint will simply
+resolve an item to its definition, and use the portability of that
+definition. In particular, when invoking trait items, the portability will be
+that of the item's *definition*, rather than that of the `impl`. See
+[Unresolved questions][unresolved] for more on this point; it's something that
+could be revised over time.
+
+#### `match_cfg`
+
+The `match_cfg` macro takes a sequence of `cfg` patterns, followed by `=>` and
+an expression. Its syntax and semantics resembles that of `match`. However,
+there are some special considerations when checking portability:
+
+* When descending into an arm of a `match_cfg`, the arm is checked against
+  portability that includes the pattern for the arm.
+
+* The portability for the `match_cfg` itself is understood as `any(p1, ...,
+  p_n)` where the `match_cfg` patterns are `p1` through `p_n`.
+
+Thus, for example, the following code will pass the lint:
+
+```rust
+#[cfg(windows)]
+fn windows_only() { .. }
+
+#[cfg(unix)]
+fn unix_only() { .. }
+
+#[cfg(any(windows, unix))]
+fn portable() {
+    // the expression here has portability `any(windows, unix)`
+    match_cfg! {
+        windows => {
+            // allowed because we are within a scope with
+            // portability `all(any(windows, unix), windows)`
+            windows_only()
+        }
+        unix => {
+            // allowed because we are within a scope with
+            // portability `all(any(windows, unix), unix)`
+            unix_only()
+        }
+    }
+}
+```
+
+If you have a `match_case` that covers *all* cases (like `windows` and
+`not(windows)`), then it imposes *no* portability constraints on its context.
+
+## The story for `std`
+
+With these basic mechanisms in hand, let's sketch out how we might apply them to
+the standard library to achieve our initial goals. This part of the RFC should
+not be considered normative; it's left to the implementation to make the final
+determination about how to set up the standard library.
+
+### The mainstream platform
+
+The "mainstream platform" will be expressed via a new primitive `cfg` pattern
+called `std`. This is the **default portability of all crates**, unless
+opted-out (see below on "subsetting `std`"). Likewise, most items in `std` will
+initially be exported at `std` portability level. These two facts together mean
+that existing uses of `std` will continue to work without issuing any warnings.
+
+The `std` portability will include several implications, e.g.:
+
+- `std` implies `any(windows, macos, linux)`
```

Comment:
99% of the platform-specific code I've wound up writing in Rust is simply `cfg(unix)` vs. `cfg(windows)`, honestly. I know that for very low-level things you can wind up having to write code that's not POSIX, but I have a hard time thinking of an example that would crop up in the standard library.